services/web/server/src/simcore_service_webserver/sharing_networks_db.py

```diff
@@ -0,0 +1,49 @@
+from aiohttp.web import Application
+from aiopg.sa import Engine
+
+from servicelib.aiohttp.application_keys import APP_DB_ENGINE_KEY
+from simcore_postgres_database.models.sharing_networks import sharing_networks
+from models_library.sharing_networks import NetworksWithAliases, SharingNetworks
+from sqlalchemy.sql import select
+from models_library.projects import ProjectID
+from sqlalchemy.dialects.postgresql import insert as pg_insert
+
+
+def _get_engine(app: Application) -> Engine:
+    engine = app.get(APP_DB_ENGINE_KEY)
+    assert engine  # nosec
+    return engine
+
+
+async def get_sharing_networks(
+    app: Application, project_id: ProjectID
+) -> SharingNetworks:
+    """Gets existing sharing_networks entry or returns an empty one"""
+
+    async with _get_engine(app).acquire() as connection:
+        query = select([sharing_networks]).where(
+            sharing_networks.c.project_uuid == f"{project_id}"
+        )
+        result = await connection.execute(query)
+        sharing_networks_row = await result.first()
+
+        if sharing_networks_row is None:
+            return SharingNetworks.create_empty(project_uuid=project_id)
+
+        return SharingNetworks.parse_obj(sharing_networks_row)
+
+
+async def update_sharing_networks(
+    app: Application, project_id: ProjectID, networks_with_aliases: NetworksWithAliases
+) -> None:
+    sharing_networks_to_insert = SharingNetworks.create(
+        project_uuid=project_id, networks_with_aliases=networks_with_aliases
+    )
+
+    async with _get_engine(app).acquire() as connection:
```

Comment:
ah finally now I understand why you needed the database...
so one question here. How is this going to work once the feature shall be accessible through the api-server as well?

so if I understand this correctly:
- User shall create a network
- webserver shall create an entry in the database
- webserver calls director-v2 to update the networks?
- director-v2 accesses the database to find what shall be connected?
- director-v2 creates the network(s)?
- director-v2 calls the dynamic-sidecar(s) to attach to the network(s)?

In the end, this whole networking is owned by the director-v2 right? That is why I asked why you would need the whole networking in a table for this...
My thought on this, is that it's a bit complex to maintain since everytime you change something you might have to change both services each time.

Reply:
There are 2 aspects to consider when dealing with  these `project_networks`:
1. the project_networks  object is already up to date and a study is being opened
2. a node is added/removed form the current project

Let's see wha happens in both situations step by step,

**Case 1.** the data regarding the project's `project_networks` object is already saved in the database:
- when starting a service which requires this information, before booting it, we need to connect the network to it
- in this case `director-v2`, before starting the service spec, it injects the networks inside the configuration
- no further action is required, service already starts and it is connected to the network.

**Case 2.** when a study was already opened, a user add/removes a node from the study or changes the study's label:
- in this case a the `update_from_workbench` is called triggering a computation of the `project_networks` object inside the database
- since docker is pretty nice we do not require to stop and restart the services in order to attach or detach networks from them, all this can be done while the container is running. So after the `project_networks` is updated a diff of networks which need to get removed and need to get attached is computed.
- Using the previous(existing) `project_networks` configuration and the one from the coming from the last workbench a list of networks which need to be detached and a list of networks which need to be attached is computed by the webserver.
- fist all `networks to be removed` are awaited together (done by the webserver)
- lastly all `networks to be attached` are awaited together (done by the webserver)
- Note: attaching and detaching networks is called by invoking director-v2's endpoint which in turns calls one on the dynamic-sidecar.

The database `project_networks` table is accessed only by director-v2 and the webserver in two different occasions:
- `webserver`: each time the workbench changes the `project_networks` entry for the project is updated
- `director-v2`: each time a new dynamic-sidecar service is started it fetches the `project_networks` from the database and recovers the list of networks it needs to the service.

I don't understand why the `api-server` might want to have access to this. From what I understood only dynamic services are the can access these project networks.

Reply:
well. until the computational services need it too.